src/mewbot/tools/dev_venv.py

```diff
@@ -0,0 +1,632 @@
+    except IndexError as exp:
+        raise NotImplementedError(
+            "You need to provide a path to a location to build the venv in"
+        ) from exp
+
+    target_dir = os.path.abspath(target_dir)
+    if not os.path.exists(target_dir):
+        raise NotImplementedError(f"{target_dir = } must exist.")
+
+    venv_name = sys.argv[2]
+    assert (
+        re.match(r"^[a-zA-Z0-9_]+$", venv_name) is not None
+    ), f"{venv_name = } failed safety checks - please only use a-z,A-Z,0-9 and _"
+
+    venv_path = os.path.join(target_dir, venv_name)
+    assert not os.path.exists(
+        venv_path
+    ), f"Cannot generate venv at {venv_path = } - there is something there."
+
+    return venv_path
+
+
+def _check_for_egg_info_folders(src_path: str, plugins_path: str) -> list[str]:
+    """
+    As part of the setup.py run process some egg-info folders are generated - check for them.
+
+    If they exist, they will be overwritten by this process.
+    Which may break other things.
+    (In particular, the python install which generated them may no longer be able to register
+    mewbot as a valid, editable package).
+    Our solution, to prevent conflicts (and to allow mypy to work - as they register as folders
+    without .py files to it, which causes it to fail) is to (eventually) move them into the root
+    folder of the venv for which they have been created - here the venv we're generating.
+
+    :param src_path:
+    :param plugins_path:
+    :return:
+    """
+    print(f"Checking {src_path} and {plugins_path} for egg-info folders.")
+
+    bad_paths: list[str] = []
+
+    for cand_path in [src_path, plugins_path]:
+        obj_list = os.listdir(cand_path)
+
+        for obj_name in obj_list:
+            cand_obj = os.path.join(cand_path, obj_name)
+
+            if not os.path.isdir(cand_obj):
+                continue
+
+            if cand_obj.endswith(".egg-info"):
+                bad_paths.append(cand_obj)
+
+    return bad_paths
+
+
+def _build_venv(venv_path: str) -> None:
+    """
+    Execute the commands needed to build the venv.
+
+    :return:
+    """
+    current_exec = sys.executable
+    print(f"Building venv using {current_exec = } in {venv_path = }")
+    subprocess.run([current_exec, "-m", "venv", venv_path], check=True)
+
+
+def _validate_venv(venv_path: str) -> str:
+    """
+    Validate the venv and return the path to the activation script.
+
+    :param venv_path:
+    :return:
+    """
+    if platform.system() != "Windows":
+        activation_path = os.path.join(venv_path, "bin", "activate")
+    else:
+        activation_path = os.path.join(venv_path, "Scripts", "activate")
+
+    assert os.path.isfile(activation_path), f"{activation_path = } not found in venv"
+    print(f"venv created - to activate using {activation_path = }")
+
+    subprocess.run(
+        [
+            activation_path,
+        ],
+        shell=True,
+        check=True,
+    )
+
+    return activation_path
```

Comment:
Fun fact: this does not work on Linux -- because activate is meant to be sourced _into_ the current shell.
Specifically it fails because the file is not executable.

Given that it only needs to set a couple of environment variables, might be easier to just mess with `os.environ`

Reply:
I have mananged to make it _runnable_ but not _functional_ -- nothing ever gets installed.
[fixes.patch.txt](https://github.com/mewbotorg/mewbot/files/11778847/fixes.patch.txt)


```
Using MEWBOT_REPO_PATH = '/home/benedict/src/javajawa/mewbot'
With TOOLS_PATH = '/home/benedict/src/javajawa/mewbot/tools'
==== 0 - Validating Input ======================================================
Checking /home/benedict/src/javajawa/mewbot/src and /home/benedict/src/javajawa/mewbot/plugins for egg-info folders.

==== 1 - Building Venv =========================================================
Building venv using current_exec = '/home/benedict/src/javajawa/mewbot/venv/bin/python' in venv_path = '/home/benedict/src/javajawa/mewbot/venv_branch'

==== 2 - Validate Venv =========================================================
venv created - attempt to activate using '. /home/benedict/src/javajawa/mewbot/venv_branch/bin/activate'

==== 3 - Update venv pip =======================================================

==== 4 - Install all deps into the venv ========================================
==================================================
Pip list in current state
==================================================

==== 5 - install the plugins ===================================================
About to attempt to install plugin_name = 'mewbot-io-discord'
About to attempt to install plugin_name = 'mewbot-discord_dice_roller'
About to attempt to install plugin_name = 'mewbot-io-file-storage'
plugin_setup_py_file = '/home/benedict/src/javajawa/mewbot/plugins/mewbot-io-file-storage/setup.py' did not exist - skipping.
==================================================
Pip list in current state
==================================================

==== 6 - Wipe pypi placeholder installs ========================================

==== 7 - Install the sub-modules in editable form ==============================
==================================================

==== 8 - Rehome egg-info folders ===============================================
plugin_src_path = '/home/benedict/src/javajawa/mewbot/plugins/mewbot-io-file-storage/src' in plugin_folder = 'mewbot-io-file-storage' did not exist.

==== 9 - Run tests on the venv =================================================



venv prepared - activate via ". /home/benedict/src/javajawa/mewbot/venv_branch/bin/activate"
```

Reply:
... mildly embaressing that.

Fixed as suggested by "activating" the venv via variable manip.